java/ql/test/library-tests/pathsanitizer/Test.java

```diff
@@ -604,4 +604,214 @@ public void fileConstructorSanitizer() throws Exception {
             sink(normalized); // $ hasTaintFlow
         }
     }
+
+    private void directoryCharsValidation(String path) throws Exception {
+        if (!path.matches("[0-9a-fA-F]{20,}")) {
+            throw new Exception();
+        }
+    }
+
+    public void directoryCharsSanitizer() throws Exception {
+        // DirectoryCharactersGuard
+        // Ensures that directory characters (/, \ and ..) cannot possibly be in the payload
+        // branch = true
+        {
+            String source = (String) source();
+            if (source.matches("[0-9a-fA-F]{20,}")) {
+                sink(source); // Safe
+            } else {
+                sink(source); // $ hasTaintFlow
+            }
+        }
+        {
+            String source = (String) source();
+            if (source.matches("[0-9a-fA-F]*")) {
+                sink(source); // Safe
+            } else {
+                sink(source); // $ hasTaintFlow
+            }
+        }
+        {
+            String source = (String) source();
+            if (source.matches("[0-9a-fA-F]+")) {
+                sink(source); // Safe
+            } else {
+                sink(source); // $ hasTaintFlow
+            }
+        }
+        {
+            String source = (String) source();
+            if (source.matches("[0-9a-fA-F\\.]+")) {
+                sink(source); // $ hasTaintFlow
+            } else {
+                sink(source); // $ hasTaintFlow
+            }
+        }
+        {
+            String source = (String) source();
+            if (source.matches("[0-9a-fA-F/]+")) {
+                sink(source); // $ hasTaintFlow
+            } else {
+                sink(source); // $ hasTaintFlow
+            }
+        }
+        {
+            String source = (String) source();
+            if (source.matches("[0-9a-fA-F\\\\]+")) {
+                sink(source); // $ hasTaintFlow
+            } else {
+                sink(source); // $ hasTaintFlow
+            }
+        }
+        {
+            String source = (String) source();
+            // exclude '.', '/', '\'
+            if (source.matches("[^0-9./\\\\a-f]+")) {
+                sink(source); // Safe
+            } else {
+                sink(source); // $ hasTaintFlow
+            }
+        }
+        {
+            String source = (String) source();
+            // '/' is not excluded
+            if (source.matches("[^0-9.\\\\a-f]+")) {
+                sink(source); // $ hasTaintFlow
+            } else {
+                sink(source); // $ hasTaintFlow
+            }
+        }
+        // branch = false
+        {
+            String source = (String) source();
+            if (source.matches("[\\./\\\\]+")) {
+                sink(source); // $ hasTaintFlow
+            } else {
+                sink(source); // $ Safe
```

Comment:
This seems wrong to me. Just because a string isn't entirely made up of those three characters, doesn't mean it is safe. Surely "../../a" isn't safe but gets to this line? (I believe that `matches` matches the regex against the whole string - it isn't reporting whether some substring matches the regex.)

Reply:
Fixed in https://github.com/github/codeql/pull/18646/commits/0d2e9ae4691c7fa9b85a112dcab8cb74a9754cef.